### PR TITLE
METRICS SDK - Calling Observable Instruments callback during metrics collection

### LINKED ISSUE #1550: Nothing calls ObservableRegistry::Observe()
I've been trying to use async instruments, but got slightly confused about the API and the status of the implementation.

The problem I ran into is that the callbacks are not getting evaluated and no data is exported. I see that the API has recently changed in https://github.com/open-telemetry/opentelemetry-cpp/pull/1388 that introduced `ObservableRegistry`. It seems like the `ObservableRegistry::Observe()` method would be the one evaluating all the registered callbacks, but it doesn't get called anywhere. According to the corresponding issue (https://github.com/open-telemetry/opentelemetry-cpp/issues/1338), all the callbacks should be evaluated before/during `collect()`.

Am I missing something here? I don't see any open issues or PRs for async instruments. If this is a bug, I'd glad to help out with a PR after some explanation about where and how this should happen.

Thanks a lot in advance!

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1550, #1556

## Changes

Please provide a brief description of the changes here.
1. Create ObservableRegistry during Meter creation.
2. Store the reference to ObservableRegistry while Instruments created from that Meter.
3. During the collection (in `Meter::Collect()`), call all the registered callbacks once.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by esigo** [comment]
is this finished?

**Comment by lalitb** [comment]
> is this finished?

Yes changes are fine, as I tested with ostream exporter, I need to add unit test. Will be finishing it today.

**Comment by lalitb** [comment]
> > is this finished?
> 
> Yes changes are fine, as I tested with ostream exporter, I need to add unit test. Will be finishing it today.

I have now added basic tests to validate the async flow, it should be good to review now.

File: sdk/test/metrics/meter_test.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+void asyc_generate_measurements(opentelemetry::metrics::ObserverResult observer, void *state)
+{
+  auto observer_long =
+      nostd::get<nostd::shared_ptr<opentelemetry::metrics::ObserverResultT<long>>>(observer);
+  observer_long->Observe(10l);
+}
+
+TEST(MeterTest, BasicAsyncTests)
+{
+  MetricReader *metricReaderPtr = nullptr;
```
Comment (esigo) [inline sdk/test/metrics/meter_test.cc:54]: nit: could we use `metric_reader_ptr` istead?

**Review by esigo — APPROVED** [review]
LGTM
Thanks for the fix.

File: sdk/test/metrics/meter_test.cc (hunk 1 above)
Comment (lalitb) [reply sdk/test/metrics/meter_test.cc]: fixed. thanks for reviewing.

File: sdk/test/metrics/meter_test.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+class MockMetricReader : public MetricReader
+{
+public:
+  MockMetricReader() = default;
+  AggregationTemporality GetAggregationTemporality(
+      InstrumentType instrument_type) const noexcept override
+  {
+    return AggregationTemporality::kCumulative;
+  }
+  virtual bool OnShutDown(std::chrono::microseconds timeout) noexcept override { return true; }
```
Comment (ThomsonTan) [inline sdk/test/metrics/meter_test.cc:26]: nit, the virtual keywords are not needed.

Comment (lalitb) [reply sdk/test/metrics/meter_test.cc]: Thanks fixed now

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**